components/layout/flex.rs

```diff
@@ -69,22 +72,192 @@ impl AxisSize {
+
+                let adjustment = match self.style.get_position().box_sizing {
+                    box_sizing::T::content_box => Au(0),
+                    box_sizing::T::border_box =>
+                        block.fragment.border_padding.inline_start_end()
+                };
+                let content_size = block.base.intrinsic_inline_sizes.preferred_inline_size
+                    - block.fragment.surrounding_intrinsic_inline_size() + adjustment;
+                self.base_size = basis.specified_or_default(content_size);
+                self.max_size = specified_or_none(self.style.max_inline_size(), containing_length)
+                    .unwrap_or(MAX_AU);
+                self.min_size = specified(self.style.min_inline_size(), containing_length);
+            },
+            Mode::Block => {
+                let basis = from_flex_basis(self.style.get_position().flex_basis,
+                                            self.style.content_block_size(),
+                                            Some(containing_length));
+                let content_size = match self.style.get_position().box_sizing {
+                    box_sizing::T::border_box => block.fragment.border_box.size.block,
+                    box_sizing::T::content_box => block.fragment.border_box.size.block
+                        - block.fragment.border_padding.block_start_end(),
+                };
+                self.base_size = basis.specified_or_default(content_size);
+                self.max_size = specified_or_none(self.style.max_block_size(), containing_length)
+                    .unwrap_or(MAX_AU);
+                self.min_size = specified(self.style.min_block_size(), containing_length);
+            }
+        }
+    }
+
+    /// Return the outer main size of the item, including paddings and margins,
+    /// clamped by max and min size.
+    pub fn outer_main_size(&self, mode: Mode) -> Au {
+        let ref fragment = self.flow.as_block().fragment;
+        let adjustment = match mode {
+            Mode::Inline => {
+                match self.style.get_position().box_sizing {
+                    box_sizing::T::content_box =>
+                        fragment.border_padding.inline_start_end() + fragment.margin.inline_start_end(),
+                    box_sizing::T::border_box =>
+                        fragment.margin.inline_start_end()
+                }
+            },
+            Mode::Block => {
+                match self.style.get_position().box_sizing {
+                    box_sizing::T::content_box =>
+                        fragment.border_padding.block_start_end() + fragment.margin.block_start_end(),
+                    box_sizing::T::border_box =>
+                        fragment.margin.block_start_end()
+                }
+            }
+        };
+        max(self.min_size, min(self.base_size, self.max_size)) + adjustment
+    }
+
+    pub fn auto_margin_num(&self, mode: Mode) -> i32 {
```

Comment:
nit: How about `auto_margin_count`? Could possibly add a comment saying that this counts the number of margins in the provided direction.